client/state/utils.js

```diff
@@ -215,6 +215,25 @@ function getInitialState( reducer ) {
 	return reducer( undefined, { type: '@@calypso/INIT' } );
 }
 
+function isValidSerializedState( schema, reducer, state ) {
+	if ( state === undefined ) {
+		return false;
+	}
+
+	// The stored state is often equal to initial state of the reducer. Because initial state
+	// is always valid, we can validate much faster by just comparing the two states. The full
+	// JSON schema check is much slower and we do it only on nontrivial states.
+	// Note that we need to serialize the initial state to make a correct check. For reducers
+	// with custom persistence, the initial state can be arbitrary non-serializable object. We
+	// need to compare two serialized objects.
```

Comment:
This is a great and valuable comment. I like how it explains the unexpected equality check before validation so someone doesn’t think they should remove it